include/libhal-soft/bit_bang_spi.hpp

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024 Khalil Estell
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <libhal/input_pin.hpp>
+#include <libhal/output_pin.hpp>
+#include <libhal/spi.hpp>
+#include <libhal/steady_clock.hpp>
+
+namespace hal::soft {
+/**
+ * @brief A bit bang implementation for spi.
+ *
+ * This implementation of spi only needs 2 hal::output_pins for sck and copi, a
+ * hal::input_pin for cipo, and a steady_clock to work.
+ */
+class bit_bang_spi : public spi
+{
+public:
+  struct pins
+  {
+    output_pin* sck;
+    output_pin* copi;
+    input_pin* cipo;
+  };
+
+  enum class delay_mode
+  {
+    with,
+    without
+  };
+
+  /**
+   * @brief Construct a new bit bang spi object
+   *
+   * @param p_pins named structure that contains pointers to the sck, copi, and
+   * cipo to be used inside of the driver
+   * @param p_steady_clock the steady clock that will be used for timing the sck
+   * line.
+   * @param p_settings the initial settings of the spi bus
+   */
+  bit_bang_spi(pins const& p_pins,
+               steady_clock& p_steady_clock,
+               settings const& p_settings = {},
+               delay_mode p_delay_mode = delay_mode::with);
```

Comment:
Missing docs for p_delay_mod

Reply:
Done, unless the wording isn't what you had in mind